core/src/libtx/proof.rs

```diff
@@ -55,35 +70,652 @@ pub fn verify(
+}
+
+/// The new, more flexible proof builder
+pub struct ProofBuilder<'a, K>
+where
+	K: Keychain,
+{
+	keychain: &'a K,
+	rewind_hash: Vec<u8>,
+	private_hash: Vec<u8>,
+}
+
+impl<'a, K> ProofBuilder<'a, K>
+where
+	K: Keychain,
+{
+	/// Creates a new instance of this proof builder
+	pub fn new(keychain: &'a K) -> Self {
+		let private_root_key = keychain
+			.derive_key(0, &K::root_key_id(), &SwitchCommitmentType::None)
+			.unwrap();
+
+		let private_hash = blake2b(32, &[], &private_root_key.0).as_bytes().to_vec();
+
+		let public_root_key = keychain
+			.public_root_key()
+			.serialize_vec(keychain.secp(), true);
+		let rewind_hash = blake2b(32, &[], &public_root_key[..]).as_bytes().to_vec();
+
+		Self {
+			keychain,
+			rewind_hash,
+			private_hash,
+		}
+	}
+
+	fn nonce(&self, commit: &Commitment, private: bool) -> Result<SecretKey, Error> {
+		let hash = if private {
+			&self.private_hash
+		} else {
+			&self.rewind_hash
+		};
+		let res = blake2b(32, &commit.0, hash);
+		SecretKey::from_slice(self.keychain.secp(), res.as_bytes()).map_err(|e| {
+			ErrorKind::RangeProof(format!("Unable to create nonce: {:?}", e).to_string()).into()
+		})
+	}
+}
+
+impl<'a, K> ProofBuild for ProofBuilder<'a, K>
+where
+	K: Keychain,
+{
+	fn rewind_nonce(&self, _secp: &Secp256k1, commit: &Commitment) -> Result<SecretKey, Error> {
+		self.nonce(commit, false)
+	}
+
+	fn private_nonce(&self, _secp: &Secp256k1, commit: &Commitment) -> Result<SecretKey, Error> {
+		self.nonce(commit, true)
+	}
+
+	/// Message bytes:
+	///     0: reserved for future use
+	///     1: wallet type (0 for standard)
+	///     2: switch commitment type
+	///     3: path depth
+	///  4-19: derivation path
+	fn proof_message(
+		&self,
+		_secp: &Secp256k1,
+		id: &Identifier,
+		switch: &SwitchCommitmentType,
+	) -> Result<ProofMessage, Error> {
+		let mut msg = [0; 20];
+		msg[2] = u8::from(switch);
+		let id_bytes = id.to_bytes();
+		for i in 0..17 {
+			msg[i + 3] = id_bytes[i];
+		}
+		Ok(ProofMessage::from_bytes(&msg))
+	}
+
+	fn check_output(
+		&self,
+		_secp: &Secp256k1,
+		commit: &Commitment,
+		amount: u64,
+		message: ProofMessage,
+	) -> Result<Option<(Identifier, SwitchCommitmentType)>, Error> {
+		if message.len() != 20 {
+			return Ok(None);
+		}
+		let msg = message.as_bytes();
+		let exp: [u8; 2] = [0; 2];
+		if msg[..2] != exp {
+			return Ok(None);
+		}
+		let switch = match SwitchCommitmentType::try_from(msg[2]) {
+			Ok(s) => s,
+			Err(_) => return Ok(None),
+		};
+		let depth = u8::min(msg[3], 4);
+		let id = Identifier::from_serialized_path(depth, &msg[4..]);
+
+		let commit_exp = self.keychain.commit(amount, &id, &switch)?;
+		match commit == &commit_exp {
+			true => Ok(Some((id, switch))),
+			false => Ok(None),
+		}
+	}
+}
+
+/// The legacy proof builder, used before the first hard fork
+pub struct LegacyProofBuilder<'a, K>
+where
+	K: Keychain,
+{
+	keychain: &'a K,
+	root_hash: Vec<u8>,
```

Comment:
Stealing this is as effective as stealing the wallet seed. Make sure you erase the memory when you're done with it. "Drop" in rust, I believe.